cpu/fe310/thread_arch.c

```diff
@@ -178,13 +178,23 @@ void cpu_switch_context_exit(void)
     UNREACHABLE();
 }
 
-void thread_yield_higher(void)
+static inline void _ecall_dispatch(uint32_t num, void *ctx)
 {
-    /* Use SW intr to schedule context switch */
-    CLINT_REG(CLINT_MSIP) = 1;
+    /* function arguments are in a0 and a1 as per ABI */
+    __asm__ volatile (
+    "mv a0, %[num] \n"
+    "mv a1, %[ctx] \n"
+    "ECALL\n"
+    "ret"
+    :
+    : [num] "r" (num), [ctx] "r" (ctx)
+    :
```

Comment:
```suggestion
    : "memory"
```

Without a memory barrier, the compiler might move stores and loads across a call to `thread_yield_higher()`. (At least with `LTO=1`.)

Reply:
Thanks, added